### PR TITLE
feat(native): Support basic writing to iceberg table

### PR DESCRIPTION
**Pull request by PingLiuPing**
## Description

Velox PR https://github.com/facebookincubator/velox/pull/14723

This commit adds support for basic Iceberg table insertion
in Prestissimo (native execution) and provides comprehensive test coverage.
All tests use the native query runner to verify Prestissimo's Iceberg
functionality

Basic insertion does not include partition transform, metrics, sort order.

## Contributor checklist

- [ ] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.
- [ ] If adding new dependencies, verified they have an [OpenSSF Scorecard](https://securityscorecards.dev/#the-checks) score of 5.0 or higher (or obtained explicit TSC approval for lower scores).

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

Prestissimo (Native Execution) Changes
* Add support basic insertion to iceberg tables.
```

### CODE REVIEW

File: presto-native-execution/presto_cpp/main/connectors/PrestoToVeloxConnector.cpp [hunk 1]
```diff
@@ -1004,6 +984,26 @@ connector::hive::HiveColumnHandle::ColumnType toHiveColumnType(
   }
 }
 
+velox::common::CompressionKind toFileCompressionKind(
```
Comment (yingsu00) [inline presto-native-execution/presto_cpp/main/connectors/PrestoToVeloxConnector.cpp]: Move this function to behind toHiveTableHandle() to keep it in the same order of the .h file.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/iceberg/TestInsertFromTpch.java [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+import org.testng.annotations.Test;
+
+import static com.facebook.presto.nativeworker.PrestoNativeQueryRunnerUtils.ICEBERG_DEFAULT_STORAGE_FORMAT;
+import static com.facebook.presto.nativeworker.PrestoNativeQueryRunnerUtils.javaIcebergQueryRunnerBuilder;
+import static com.facebook.presto.nativeworker.PrestoNativeQueryRunnerUtils.nativeIcebergQueryRunnerBuilder;
+
+public class TestInsertFromTpch
+        extends AbstractTestQueryFramework
+{
+    private static final String TPCH_SCHEMA = "iceberg.tpch";
```
Comment (yingsu00) [inline presto-native-execution/src/test/java/com/facebook/presto/nativeworker/iceberg/TestInsertFromTpch.java:29]: nit: add an empty line below

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/iceberg/TestInsertFromTpch.java [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        try {
+            assertUpdate(String.format("CREATE TABLE %s (" +
+                    "nationkey BIGINT, " +
+                    "name VARCHAR, " +
+                    "regionkey BIGINT, " +
+                    "comment VARCHAR" +
+                    ") WITH (format = 'PARQUET')", targetTable));
+
+            long rowCount = (Long) computeActual(String.format("SELECT COUNT(*) FROM %s", sourceTable)).getOnlyValue();
+            assertUpdate(String.format("INSERT INTO %s SELECT * FROM %s", targetTable, sourceTable), rowCount);
```
Comment (yingsu00) [inline presto-native-execution/src/test/java/com/facebook/presto/nativeworker/iceberg/TestInsertFromTpch.java]: It would be nice if we can do multiple rounds of insertion tests on the nation or region tables.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/iceberg/TestInsertFromTpch.java [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+        try {
+            assertUpdate(String.format("CREATE TABLE %s (" +
+                    "custkey BIGINT, " +
+                    "name VARCHAR, " +
+                    "nationkey BIGINT, " +
+                    "acctbal DOUBLE" +
+                    ") WITH (format = 'PARQUET')", targetTable));
+
+            long rowCount = (Long) computeActual(String.format("SELECT COUNT(*) FROM %s WHERE acctbal > 0", sourceTable)).getOnlyValue();
+            assertUpdate(String.format("INSERT INTO %s " +
```
Comment (yingsu00) [inline presto-native-execution/src/test/java/com/facebook/presto/nativeworker/iceberg/TestInsertFromTpch.java:148]: Add a test case where filter is evaluated as false and 0 rows being inserted?

**Review by yingsu00 — COMMENTED** [review]
@PingLiuPing Mostly look good. The first commit needs to come with tests. Would it be possible to move the TestUnpartitionedWrite and TestInsertFromTpch from the second commit to the first commit?

File: presto-native-execution/presto_cpp/main/connectors/PrestoToVeloxConnector.cpp (hunk 1 above)
Comment (PingLiuPing) [reply presto-native-execution/presto_cpp/main/connectors/PrestoToVeloxConnector.cpp]: Thanks.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/iceberg/TestInsertFromTpch.java (hunk 2 above)
Comment (PingLiuPing) [reply presto-native-execution/src/test/java/com/facebook/presto/nativeworker/iceberg/TestInsertFromTpch.java:29]: Thanks.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/iceberg/TestInsertFromTpch.java (hunk 3 above)
Comment (PingLiuPing) [reply presto-native-execution/src/test/java/com/facebook/presto/nativeworker/iceberg/TestInsertFromTpch.java]: Yes, good idea.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/iceberg/TestInsertFromTpch.java (hunk 4 above)
Comment (PingLiuPing) [reply presto-native-execution/src/test/java/com/facebook/presto/nativeworker/iceberg/TestInsertFromTpch.java:148]: Yes, I missed this edge case.

**Comment by PingLiuPing** [comment]
@yingsu00  Thanks for the review comments. I have fixed them and re-arranged the commit. 

For the macos build error, that is an issue in prestissimo macos build script, I have submitted a PR to fix it. And now it should pass the CI. I have rebased the code.

**Review by yingsu00 — APPROVED** [review]

File: presto-native-execution/presto_cpp/main/connectors/IcebergPrestoToVeloxConnector.cpp [hunk 5]
```diff
@@ -274,4 +275,74 @@ IcebergPrestoToVeloxConnector::createConnectorProtocol() const {
...
+      insertHandle->handle.connectorHandle->_type);
+
+  const auto inputColumns =
+      toHiveColumns(icebergInsertTableHandle->inputColumns, typeParser);
+
+  return std::make_unique<
+      velox::connector::hive::iceberg::IcebergInsertTableHandle>(
+      inputColumns,
+      std::make_shared<velox::connector::hive::LocationHandle>(
+          fmt::format("{}/data", icebergInsertTableHandle->outputPath),
```
Comment (aditi-pandit) [inline presto-native-execution/presto_cpp/main/connectors/IcebergPrestoToVeloxConnector.cpp:326]: fmt::format("{}/data", icebergInsertTableHandle->outputPath) seems repeated at multiple places in the code. Abstract an inline function for this and reuse.

File: presto-native-execution/presto_cpp/main/connectors/PrestoToVeloxConnector.h [hunk 6]
```diff
@@ -52,6 +52,9 @@ std::unique_ptr<velox::connector::ConnectorTableHandle> toHiveTableHandle(
     const VeloxExprConverter& exprConverter,
     const TypeParser& typeParser);
 
+velox::common::CompressionKind toFileCompressionKind(
```
Comment (aditi-pandit) [inline presto-native-execution/presto_cpp/main/connectors/PrestoToVeloxConnector.h:55]: If you are separating a HiveToPrestoVeloxConnector file then this can be moved there as well.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/iceberg/TestSelect.java
Comment (aditi-pandit) [inline presto-native-execution/src/test/java/com/facebook/presto/nativeworker/iceberg/TestSelect.java:1]: @PingLiuPing : We have been adding native tests to presto-native-tests modules so that presto-native-execution remains light-weight and has only basic tests that Meta and us care about. 

Would you be able to add some tests into presto-native-tests, and also enhance the framework so that we can use IcebergQueryRunner it in instead of always Hive, along with side-car as that is more realistic setup. https://github.com/prestodb/presto/tree/master/presto-native-tests/src/test/java/com/facebook/presto/nativetests

**Review by aditi-pandit — COMMENTED** [review]
@PingLiuPing Sorry for few late comments. But you can add in follow up PRs.

**Comment by psantos-denodo** [comment]
Hi @PingLiuPing,

Could you clarify what is the limitation regarding sort order? Could you include an example? Thanks!
